### ISSUE TITLE
Private stack sharing with Dockerhub Encrypted Storage

### ISSUE THREAD

**Issue by seertenedos** [Taisun-Docker/taisun#1]
It would be great if there was a way to have a private stack repository. Then the app could show the public ones and private.

This could work even if it is just a folder/volume outside the docker image/app though being able to use a private git repository may be better for versioning and being accessible in case you need to rebuild your docker environment.

The main reason for private stacks is for having a heap of stacks with pre configured defaults based on personal setups to speed up rebuilds or system/docker upgrades.

I currently use docker compose files and do it all manually but your UI looks like it would make that all so much easier

**Comment by thelamer** [comment]
I have actually been thinking about stuff like this lately. 
Initial implementation idea was to store encrypted blobs from the user , but that could be abused too easily for storing illicit content or just files in general on the projects dime . 

A solution like this needs to be secure , but also needs to push the hosting costs off of Taisun . 

Would it be acceptable to implement a system like this where you as a user would need to enter your docker hub credentials in app to push or pull your stored compose templates ?
The encryption for the file could be based on combining your username and password into an unlock key on the download side of it .

**Comment by seertenedos** [comment]
I did not realise you could store them on dockerhub. If they are a private
store then should be fine. Another option is Google drive or Bitbucket or
any other data store that has a free private store.

It may be better to let people pick a password or encryption key if they
want as one issue I see is if the docker hub password changes you could not
access the files again.

Lastly I would make it sort of like a plugin in the code so you could have
a local store on a bind mount or volume or online storage options. S3 is
another option since you can self host for free or use AWS for free for a
year. I love the idea of it been online but would be happy with a local
store I can backup till it is sorted.

That said if docker hub does all this it is cool. Only advantage of other
stores is you could store other things like config if you wanted.

**Comment by thelamer** [comment]
At it's core this application has no concept of local storage it uses local Docker metadata to show everything in the interface. 

As compose files are tiny I do not see compelling reason to store them in local docker volumes locally as the primary reason for storing this kind of data is when you need to do a new setup or you lost something locally. 

Dockerhub is just tarball layers , you can copy any arbitrary data into a scratch base. People do this for things like build artifacts all the time. 

I see your point regarding using Hub auth credentials for encryption and decryption, separating the keys out also provides users the ability to share the endpoints as they could send a blob to their buddies in a format like this: 

user/metadataendpoint:tag|decryptionkeyhash

The user could then put this in their local taisun install and they would be presented with the "private" web form to launch the custom template. 

I cannot promise an ETA on something like this, but template sharing is definitely at the core of what this is all for, I made the bar to uploading to https://stacks.taisun.io/ as low as possible, just need a github account and you can publicly post to everyone using Taisun. 

We should have a private share option.

**Comment by thelamer** [comment]
Sorry forgot to mention the files that would be encrypted and stored would be in this format: 

https://github.com/Taisun-Docker/taisun/wiki/Templates

**Comment by seertenedos** [comment]
On second thoughts you could use docker image but use one of the free sites
that let you store private images. I have used one before but need to go
though my notes to find the name.

If you went that way it you could just make it a private version of the
current site you have for them that you bring up in another docker
container locally so very little code would need to change.

**Comment by thelamer** [comment]
@seertenedos you can try out the initial implementation here: 
```
docker run -d --name=taisun \
-v /var/run/docker.sock:/var/run/docker.sock \
-p 3000:3000 \
-p 8000:8000 \
taisun/webapp:latest.dev
```
You will see an upload private in the stacks menu. 
The design is to allow the user to enter a custom stack (to test you can just grab one under details here https://stacks.taisun.io/) . 
This stack is encrypted using a random GUID as the password and added to the gpg base image. 

What you end up with is a combination of the endpoint and a password to decrypt the file in the container IE here is one I did for Plex: (under browse stacks new option for From Endpoint)

taisun/test-data:plex|3ad2e8a7-95fd-4a8d-b1eb-0d24fd7d2221

You can store these strings somewhere safe or share them with other users. 

As for making what you originally requested, you have to understand that the central infrastructure for this application is all serverless. Making some kind of local endpoint that acts the same as the stacks.taisun.io endpoint while not hard, would require the user to run a local database backend along with the custom frontend api. In my experience working support for Linuxserver.io, this is a huge headache to support on remote users systems as they do not expect to be system admins and own their own infrastructure , they expect for shit to just work. 

So in summary I did these changes, because the system needs a sharing mechanism outside of my centralized shit and it allows a normal user that has only signed up for dockerhub to securely store a custom stack config even containing secrets remotely and they only have to store a string somewhere safe to relaunch it anywhere. 

I need to vet this locally and I will push to the linuxserver endpoint, probably within the next 24 hours.

**Comment by thelamer** [comment]
As this has been released and will be ingested by Linuxserver I am closing this issue.
